registry/storage/driver/s3-aws/s3.go

```diff
@@ -665,103 +665,104 @@ func (d *driver) Reader(ctx context.Context, path string, offset int64) (io.Read
+	res, err := d.S3.CreateMultipartUploadWithContext(ctx, &s3.CreateMultipartUploadInput{
+		Bucket:               aws.String(d.Bucket),
+		Key:                  aws.String(key),
+		ContentType:          d.getContentType(),
+		ACL:                  d.getACL(),
+		ServerSideEncryption: d.getEncryptionMode(),
+		SSEKMSKeyId:          d.getSSEKMSKeyID(),
+		StorageClass:         d.getStorageClass(),
+	})
+	if err != nil {
+		return nil, err
+	}
+	return res.UploadId, nil
+}
 
-		var allParts []*s3.Part
-		for _, multi := range resp.Uploads {
-			if key != *multi.Key {
-				continue
-			}
+func (d *driver) inProgressUpload(ctx context.Context, path string) (uploadID *string, err error) {
+	var empty bool
+	if err := d.S3.ListMultipartUploadsPagesWithContext(ctx, &s3.ListMultipartUploadsInput{
+		Bucket: aws.String(d.Bucket),
+		Prefix: aws.String(d.s3Path(path)),
+	}, func(page *s3.ListMultipartUploadsOutput, lastPage bool) bool {
+		if len(page.Uploads) == 0 {
+			empty = true
+			return false
+		}
 
-			partsList, err := d.S3.ListPartsWithContext(ctx, &s3.ListPartsInput{
-				Bucket:   aws.String(d.Bucket),
-				Key:      aws.String(key),
-				UploadId: multi.UploadId,
-			})
-			if err != nil {
-				return nil, parseError(path, err)
-			}
-			allParts = append(allParts, partsList.Parts...)
-			for *partsList.IsTruncated {
-				partsList, err = d.S3.ListPartsWithContext(ctx, &s3.ListPartsInput{
-					Bucket:           aws.String(d.Bucket),
-					Key:              aws.String(key),
-					UploadId:         multi.UploadId,
-					PartNumberMarker: partsList.NextPartNumberMarker,
-				})
-				if err != nil {
-					return nil, parseError(path, err)
-				}
-				allParts = append(allParts, partsList.Parts...)
+		for _, upload := range page.Uploads {
+			if *upload.Key == d.s3Path(path) {
+				uploadID = upload.UploadId
+				return false
 			}
-			return d.newWriter(ctx, key, *multi.UploadId, allParts), nil
 		}
+		return true
+	}); err != nil {
+		return nil, fmt.Errorf("list multipart uploads pages: %w", err)
+	}
 
-		// resp.NextUploadIdMarker must have at least one element or we would have returned not found
-		listMultipartUploadsInput.UploadIdMarker = resp.NextUploadIdMarker
+	if !empty {
+		return uploadID, nil
+	}
 
-		// from the s3 api docs, IsTruncated "specifies whether (true) or not (false) all of the results were returned"
-		// if everything has been returned, break
-		if resp.IsTruncated == nil || !*resp.IsTruncated {
-			break
+	fi, err := d.Stat(ctx, path)
+	if err != nil {
+		return nil, parseError(path, err)
+	}
+
+	if fi.Size() > 0 {
+		return nil, storagedriver.Error{
+			DriverName: driverName,
+			Detail:     fmt.Errorf("append to non zero-size path %s unsupported", path),
 		}
 	}
-	return nil, storagedriver.PathNotFoundError{Path: path}
+	return nil, nil
+}
+
+func (d *driver) listParts(ctx context.Context, path string, uploadID *string) (parts []*s3.Part, err error) {
```

Comment:
Similar to my previous comment regarding named returns: let's avoid them if possible; saving an extra line of slice var declaration is not worth it here IMHO.